### PR TITLE
Sidebars improvement

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
Replaces the current drawer with DrawerLayout a polyfill for RN's DrawerLayoutAndroid's and also brings back the channel loader when switching channels

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-10503

### CODE REVIEW

File: app/components/sidebars/main/main_sidebar.js [hunk 1]
```diff
@@ -118,8 +109,8 @@ export default class ChannelDrawer extends Component {
     }
 
     handleAndroidBack = () => {
-        if (this.refs.drawer && this.refs.drawer.isOpened()) {
-            this.refs.drawer.close();
+        if (this.refs.drawer) {
```
Comment (koxen) [inline app/components/sidebars/main/main_sidebar.js]: removing this && part won't be going to end well
you can hook 
onDrawerClose={this.handleDrawerClose}
onDrawerOpen={this.handleDrawerOpen}
to know whether it's open though

File: app/actions/views/more_dms.js [hunk 2]
```diff
@@ -27,7 +27,7 @@ export function makeDirectChannel(otherUserId) {
             channel = result.data;
         }
 
-        if (channel) {
+        if (channel && swithToChannel) {
```
Comment (koxen) [inline app/actions/views/more_dms.js]: typo

**Comment by koxen** [comment]
also let me know if you see drawers bug out after exiting from "markdown code view" screen by using back button on android (when its bugged it never closes after for example selecting channel) cause the fix is a very non obvious one and I bet you'll see it the same as in my version

File: app/components/channel_loader/channel_loader.js [hunk 3]
```diff
@@ -9,15 +9,79 @@ import {
...
+                handleSelectChannel(channel.id);
+
+                // mark the channel as viewed after all the frame has flushed
+                markChannelAsRead(channel.id, currentChannelId);
+                if (channel.id !== currentChannelId) {
+                    markChannelAsViewed(currentChannelId);
+                }
+
+                setChannelLoading(false);
+            }, 250);
```
Comment (koxen) [inline app/components/channel_loader/channel_loader.js]: So we are adding a hardcoded delay and UX slowdown just to show the ChannelLoader? not a big fan of this solution

**Comment by enahum** [comment]
@koxen the drawers "work" after going back, but somehow the back button on the code screen is not even working for me.

I'll try and see why

**Comment by enahum** [comment]
Ok so is not bugging out as you mention after explicitly handling the back button on the code screen

**Comment by koxen** [comment]
> Ok so is not bugging out as you mention after explicitly handling the back button on the code screen

alright, my take at this was problem was something like this:

```
        // [FIXME] Using navigator.push without overriding onNavigatorEvent with close button handler
        // breaks navigation stack when code modal is closed
        // As a workaround we will use showModal like in other parts of the code
        // With the downside being different, probably longer, animation type.

        const options = {
            screen: 'Code',
            title,
            animated: true,
            animationType: 'fade',
            backButtonTitle: '',
            passProps: {
                content: this.props.content,
            },
            navigatorStyle: {
                navBarTextColor: theme.sidebarHeaderTextColor,
                navBarBackgroundColor: theme.sidebarHeaderBg,
                navBarButtonColor: theme.sidebarHeaderTextColor,
                screenBackgroundColor: theme.centerChannelBg,
            },
        });

        if (Platform.OS === 'android') {
            navigator.showModal(options);
        } else {
            navigator.push(options);
        }
```

File: app/components/channel_loader/channel_loader.js [hunk 4]
```diff
@@ -9,15 +9,79 @@ import {
...
+    static getDerivedStateFromProps(nextProps, prevState) {
+        if (!nextProps.channelIsLoading && prevState.switch) {
+            return {
+                switch: false,
+                channel: null,
+                currentChannelId: null,
+            };
+        }
+
+        return prevState;
```
Comment (hmhealey) [inline app/components/channel_loader/channel_loader.js]: I think you might be able to return null here if the state is unchanged

File: app/components/channel_loader/channel_loader.js [hunk 5]
```diff
@@ -9,15 +9,79 @@ import {
...
+    }
+
+    componentDidMount() {
+        EventEmitter.on('switch_channel', this.handleChannelSwitch);
```
Comment (hmhealey) [inline app/components/channel_loader/channel_loader.js]: Why use the event emitter for this? Couldn't we detect this by watching if the current channel ID changes?

Comment (enahum) [reply app/components/channel_loader/channel_loader.js]: yes we could but it was taking longer to execute than doing this, so this is for performance really

**Comment by enahum** [comment]
@hmhealey changes made and rebased

File: app/components/channel_loader/channel_loader.js (hunk 3 above)
Comment (sudheerDev) [reply app/components/channel_loader/channel_loader.js]: This is odd. i can live with loader the first time but a recurring channel switch does not require a loader as we do have the information.  I think we should re-consider later about loader for loaded channels.

File: app/components/sidebars/main/main_sidebar.js [hunk 6]
```diff
@@ -131,8 +125,8 @@ export default class ChannelDrawer extends Component {
     };
 
     closeChannelDrawer = () => {
-        if (this.refs.drawer && this.refs.drawer.isOpened()) {
-            this.refs.drawer.close();
+        if (this.drawerOpened) {
+            this.refs.drawer.closeDrawer();
```
Comment (sudheerDev) [inline app/components/sidebars/main/main_sidebar.js]: This causes an issue for me on team switch. 
screen shot 2018-06-25 at 7 38 04 pm

File: app/components/search_bar/search_bar.android.js [hunk 7]
```diff
@@ -92,6 +92,7 @@ export default class SearchBarAndroid extends PureComponent {
         InteractionManager.runAfterInteractions(() => {
             this.setState({
                 isFocused: false,
+                value: '',
```
Comment (sudheerDev) [inline app/components/search_bar/search_bar.android.js]: 👍

File: app/components/channel_loader/channel_loader.js [hunk 8]
```diff
@@ -9,15 +9,79 @@ import {
...
+        style: CustomPropTypes.Style,
         theme: PropTypes.object.isRequired,
     };
 
+    static defaultProps = {
+        maxRows: 6,
+    };
+
+    state = {
+        switching: false,
```
Comment (sudheerDev) [inline app/components/channel_loader/channel_loader.js]: I don't see a reference for this should this be `switch`?.

File: app/components/sidebars/main/main_sidebar.js (hunk 6 above)
Comment (koxen) [reply app/components/sidebars/main/main_sidebar.js]: yeah this part of if should be added back

File: app/components/channel_loader/channel_loader.js (hunk 3 above)
Comment (enahum) [reply app/components/channel_loader/channel_loader.js]: that's true when the channel has been previously loaded and the posts haven't been cleaned up, we can look at improving this further in any case in a later PR, feel free to investigate and submit one ;)

File: app/screens/channel/channel.js [hunk 9]
```diff
@@ -137,15 +151,41 @@ export default class Channel extends PureComponent {
         }
     };
 
-    channelDrawerRef = (ref) => {
+    channelLoaderDimensions = () => {
```
[the remark this answers is not in the document]
Comment (enahum) [reply app/screens/channel/channel.js]: have you tried rotating the device without this piece of code? this actually do the job

File: app/components/sidebars/main/main_sidebar.js (hunk 6 above)
Comment (enahum) [reply app/components/sidebars/main/main_sidebar.js]: will fix

File: app/components/channel_loader/channel_loader.js (hunk 8 above)
Comment (enahum) [reply app/components/channel_loader/channel_loader.js]: nice catch, will fix

**Comment by enahum** [comment]
@sudheerDev changes made

File: app/screens/channel/channel.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (sudheerDev) [reply app/screens/channel/channel.js]: I am 0/5 on this. It does work but it only needs to work when we load the channel. so, I thought constructor would do.

[the remark this answers is not in the document]
Comment (sudheerDev) [reply app/screens/channel/channel.js]: Never mind. This does not update all that much anyway

**Pull request merged**